### PR TITLE
remove rsmBuilder restriction on joined table inheritance schemes

### PR DESCRIPTION
**Pull request by jnfry**
See https://github.com/doctrine/orm/issues/7151#issuecomment-545289066

### CODE REVIEW

**Review by SenseException — DISMISSED** [review]
In would like to see another reviewer take a look at this PR. I'm not sure if the tests are sufficient for this change.

File: tests/Doctrine/Tests/ORM/Functional/NativeQueryTest.php [hunk 1]
```diff
@@ -370,6 +372,31 @@ public function testAbstractClassInSingleTableInheritanceSchemaWithRSMBuilderThr
         $rsm->addRootEntityFromClassMetadata(CompanyContract::class, 'c');
     }
 
+    /**
+     * @group rsm-sti
```
Comment (beberlei) [inline tests/Doctrine/Tests/ORM/Functional/NativeQueryTest.php]:
```suggestion
     * @group GH-7151
```

File: tests/Doctrine/Tests/ORM/Functional/NativeQueryTest.php [hunk 2]
```diff
@@ -370,6 +372,31 @@ public function testAbstractClassInSingleTableInheritanceSchemaWithRSMBuilderThr
...
+    {
+        $rsm = new ResultSetMappingBuilder($this->em);
+
+        $rsm->addRootEntityFromClassMetadata(CompanyAuction::class, 'c');
+
+        self::assertSame(CompanyAuction::class, $rsm->getClassName('c'));
+    }
+
+    /**
+     * @group rsm-sti
```
Comment (beberlei) [inline tests/Doctrine/Tests/ORM/Functional/NativeQueryTest.php]:
```suggestion
     * @group GH-7151
```

**Review by beberlei — CHANGES_REQUESTED** [review]
This is missing a test in `LazyCriteriaCollectionTest` that triggers the code in `ManyToManyPersister::loadCriteria` with a joined table inheritance entity example.

**Comment by chq81** [comment]
@jnfry Hey there. Thanks for your work. Did you have the possibility to look over the missing tests?

**Comment by jnfry** [comment]
Hey there @chq81, sorry for taking so long to get back to this.

I've had a look at what @beberlei is asking and am unsure what to do. The current `LazyCriteriaCollectionTest` appears to effectively cover the responsibilities of the class (ie. only calling methods on the `Persister` when necessary and caching the result). It doesn't care which type of `Persister` it wraps, and adding a test which asserts it calls `Persister::loadCriteria()` would not test the change I've made.

I agree that there likely needs to be more tests to cover this change, but the `LazyCriteriaCollection` appears to be unaffected. Did you perhaps mean a different test case @beberlei?

I think adding a PersisterTest with a join table inheritance example would be better? Though I'm not sure which of the `*PersisterTest`'s to add this to. If you could perhaps elaborate on what exactly needs to be tested that would greatly help me out. This is my first contribution to Doctrine so I'm a bit lost :sweat_smile:

**Comment by SenseException** [comment]
@jnfry The current tests are covering a `ManyToOne` relation case with `CompanyEvent` / `CompanyAuction`. Please take a look at the content of the mapping of these classes. What's now needed are `ManyToMany` cases.

**Comment by beberlei** [comment]
@jnfry your change suggests this just works, but I am not convinced. What I am missing here is a functional test of the scenario end to end. In one of the joined table inheritance tests you need to add a test case for ->matching that triggers this code.

Your tests at the moment only check small things on ResultSetMapping, but what if the contents of the RSM are wrong so that *other* parts of the system fail, mainly the persisters?

**Pull request closed without merging**